packages/graphql-client/src/graphql-client/graphql-client.ts

```diff
@@ -165,3 +184,276 @@ function generateRequest(
+                const body = chunk
+                  .slice(
+                    chunk.indexOf(HEADER_SEPARATOR) + HEADER_SEPARATOR.length,
+                  )
+                  .trim();
+                return body;
+              });
+
+            if (chunkBodies.length > 0) {
+              yield chunkBodies;
+            }
+
+            buffer = buffer.slice(lastBoundaryIndex + boundary.length);
+
+            if (buffer.trim() === `--`) {
+              buffer = "";
+            }
+          }
+        }
+      } catch (error) {
+        throw new Error(
+          `Error occured while processing stream payload - ${getErrorMessage(
+            error,
+          )}`,
+        );
+      }
+    },
+  };
+}
+
+function generateRequestStream(
+  fetch: ReturnType<typeof generateFetch>,
+): GraphQLClient["requestStream"] {
+  return async (...props) => {
+    if (!DEFER_OPERATION_REGEX.test(props[0])) {
+      throw new Error(
+        formatErrorMessage(
+          "This operation does not result in a streamable response - use request() instead.",
+        ),
+      );
+    }
+
+    try {
+      const response = await fetch(...props);
+
+      const { status, statusText } = response;
+
+      if (!response.ok) {
+        throw new Error(statusText, { cause: response });
+      }
+
+      const responseContentType = response.headers.get("content-type") || "";
+      const isNotSupportedContentType = Object.values(CONTENT_TYPES).every(
+        (type) => !responseContentType.includes(type),
+      );
+
+      if (isNotSupportedContentType) {
+        throw new Error(
+          `${UNEXPECTED_CONTENT_TYPE_ERROR} ${responseContentType}`,
+          { cause: response },
+        );
+      }
+
+      if (responseContentType.includes(CONTENT_TYPES.json)) {
+        return {
+          async *[Symbol.asyncIterator]() {
+            const processedResponse = await processJSONResponse(response);
+
+            yield {
+              ...processedResponse,
+              hasNext: false,
+            };
+          },
+        };
+      }
+
+      const boundaryHeader = (responseContentType ?? "").match(
+        BOUNDARY_HEADER_REGEX,
+      );
+      const boundary = `--${boundaryHeader ? boundaryHeader[1] : "-"}`;
+
+      if (
+        !response.body?.getReader &&
+        !(response.body as any)![Symbol.asyncIterator]
+      ) {
+        throw new Error(
+          "API multipart response did not return an iterable body",
+          { cause: response },
+        );
+      }
+
+      const streamBodyIterator = getStreamBodyIterator(response);
+
+      let combinedData: Record<string, any> = {};
+      let responseExtensions: Record<string, any> | undefined;
+
+      return {
+        async *[Symbol.asyncIterator]() {
+          try {
+            let streamHasNext = true;
+
+            for await (const chunkBodies of readStreamChunk(
+              streamBodyIterator,
+              boundary,
+            )) {
+              const dataArray: {
```

Comment:
Same thing here - I think it would improve readability if this were

```ts
const dataArray = parseChunkBodies(chunkBodies);
```

to separate the parsing logic from the processing logic.